### ISSUE TITLE
Buffer Overflow with large PDF

### ISSUE THREAD

**Issue by natewiebe13** [wkhtmltopdf/wkhtmltopdf#2093]
Generating a large PDF with a header and footer, results in too many files being opened and throws a QT error. Increasing nofile in /etc/security/limits.conf results in a buffer overflow:

```
nate@nate-desktop:~/pdf-test$ ./script.sh 
Loading pages (1/6)
Counting pages (2/6)                                               
Resolving links (4/6)                                                       
Loading headers and footers (5/6)                                           
*** buffer overflow detected ***: wkhtmltopdf terminated     ] 9%
[========= Backtrace: =========
===>  /lib/x86_64-linux-gnu/libc.so.6         /lib/x86_64-linux-gnu/libc.so.6        /lib/x86_64-linux-gnu/libc.so.6       /lib/x86_64-linux-gnu/libc.so.6  wkhtmltopdf  wkhtmltopdf  wkhtm00400000-02674000 r-xp 00000000 08:01 1310769                            /usr/local/bin/wkhtmltopdfthread.so.0        ***    ] 9%
02874000-02a7b000 r--p 02274000 08:01 1310769                            /usr/local/bin/wkhtmltopdf
02a7b000-02a94000 rw-p 0247b000 08:01 1310769                            /usr/local/bin/wkhtmltopdf
02a94000-02ac7000 rw-p 00000000 00:00 0 
03bc9000-04c44000 rw-p 00000000 00:00 0                                  [heap]
7fe2f8ee2000-7fe2f8ee3000 ---p 00000000 00:00 0 
7fe2f8ee3000-7fe2f96e3000 rw-p 00000000 00:00 0                          [stack:9336]
7fe2f96e3000-7fe2f96e4000 ---p 00000000 00:00 0 
7fe2f96e4000-7fe2f9ee4000 rw-p 00000000 00:00 0                          [stack:9335]
7fe2f9ee4000-7fe2f9ee5000 ---p 00000000 00:00 0 
7fe2f9ee5000-7fe2fa6e5000 rw-p 00000000 00:00 0                          [stack:9334]
7fe2fa6e5000-7fe2fa6e6000 ---p 00000000 00:00 0 
7fe2fa6e6000-7fe2faee6000 rw-p 00000000 00:00 0                          [stack:9333]
7fe2faee6000-7fe2faee7000 ---p 00000000 00:00 0 
7fe2faee7000-7./script.sh: line 3:  8995 Aborted                 (core dumped) wkhtmltopdf --page-size 'Letter' --no-outline --dpi '300' --encoding 'UTF-8' --margin-top '1.25in' --margin-right '0.75in' --margin-bottom '0.75in' --margin-left '0.75in' --orientation 'portrait' --zoom '1.042' --footer-html 'http://localhost/pdf/footer.html' --header-html 'http://localhost/pdf/header.html' --use-xserver test.html --disable-smart-shrinking test.pdf
```

**Comment by natewiebe13** [comment]
@ashkulz have you been able to look into this yet?

**Comment by ashkulz** [comment]
not really, can you upload a test case which can reproduce this? The stack trace points to an error in Qt itself, which I doubt is the real problem.

**Comment by yuribelan** [comment]
I think the problem is not with the amount of data, but with the number of files open at one time (as the OP originally said).  I'm trying to generate a file with 30 requests (1 html, 2 font, 27 image files), and I get a TimeoutException.  If remove one image, the file generates in 1.49 seconds.

I'm using MS Server 2008 R2 wkhtmltopdf 0.12.1

**Comment by fenilp** [comment]
Hi,

I had also got the same error of too many files open in case of large PDFs with header and footer. Thus I need to remove the headers and footers to get it worked. I have been looking for a solution or an alternative for this. Any suggestion would be welcomed. 

@rsteinwand You mentioned about merging PDFs. Does wkhtmltopdf provide merging facility?

Thanks,
Fenil

**Comment by rsteinwand** [comment]
Fenilp,

I combine the html files before converting to pdf.

**Comment by joshuapinter** [comment]
I'm having the _exact_ same issue as the OP, including upping the limit of open files allowed, which changed the error from "Too many open files." to this buffer overflow.


I'm generating simple, but long, PDFs that are roughly 500 - 1000 pages with a simple header and footer. Without the header and footer it works just fine. It seems like adding the header and footer exponentially increases the resources required to generate the PDF.

Any update to this or suggestions? 

Many Thanks.

**Comment by joshuapinter** [comment]
Just wanted to follow-up here with what worked with us. We've been tracking this down for the past week and came up with the following solution. Again, we're creating PDFs that are 1000s of pages long with both a header and a footer.

Two things allowed us to work around the issues:

1. I increased the open file and stack size limits of our servers by adding the following config to `/etc/security/limits.conf`:

  ```bash
*    hard nofile 1000000
*    soft nofile 1000000
*    soft stack  81920
root hard nofile 1000000 # Need these two lines as wildcards (above)
root soft nofile 1000000 # are not applied to root.
root soft stack  81920
```

2. Instead of using a S3 image in the header, I store the image locally. Getting a header image from S3 seemed to require more resources.

You're gonna have to play with the limits to see what works with the size of the PDFs you're generating but it's surprising how high these limits have to be to prevent issues. 

Ideally, we would want to refactor wkhtmltopdf so that it does not use as many resources for appending headers and footers to a PDF document.

Thanks for all your effort on the library.

**Comment by daniar23** [comment]
I found out that the problem was in setting up pathes of footer and header with http urls. So I've changed them to local pathes and it solved the problem.

**Comment by joshuapinter** [comment]
@daniar23  Can you give an example of what your header and footer paths looked like before and then after you changed them?

**Comment by daniar23** [comment]
@joshuapinter it was like stage.website.com/fileadmin/pdfs/footer.html, 
now /var/www/vhost/myproject/fileadmin/pdfs/footer.html

**Comment by joshuapinter** [comment]
@daniar23 Interesting... I'm using this with [wicked-pdf](https://github.com/mileszs/wicked_pdf) and it uses local paths for the header and footer but I was still bumping up against the limits until I increased them. 

Out of curiosity, how many pages were you trying to generate when it was giving you a problem?

**Comment by daniar23** [comment]
@joshuapinter I'm using snappy. But I was directly testing in linux-shell, to be sure that it's not a snappy problem. I was trying to generate 432 pages PDF. After switching header and footer to local pathes, I was even able to generate PDFs with 700 pages.

**Comment by joshuapinter** [comment]
@daniar23 Right. Okay. I was hitting those limits when I was generating PDFs with 1000s of pages long. 

The local vs remote paths is a valid one, even for images used inside of your header and footer. I originally had a remote image in my header and it severely limited how many pages I could generate. I switched that to a local image and it allowed me to generate a lot more.

It seems WKTHMLTOPDF opens up a separate thread or file or whatever for each remote path it has to fetch, which causes problems on very long PDFs.

**Comment by lorenzos** [comment]
@daniar23 Thank for your suggestion, after changing header and footer to local paths I managed to create the PDF. Unfortunately I had dynamic content in it (they were PHP files called with URL parameters): I think I'll try to set up an automatic download of that dynamic pages before generating PDF, so then I can use local paths to downloaded HTML files.

**Comment by joshuapinter** [comment]
I'm coming back here two years later. We've hit some pretty big scale at our company so our servers have increased in size, including moving to bare-metal servers at http://packet.net. 

We were bumping up against the `ulimit -n #=> 1024` when generating large PDFs with images and links, but we only needed to increase the limit to 2048 (`ulimit -n 2048`) and we were able to generate PDFs up to 1000 pages without a problem.

I think the size of the server plays a huge role in what these limits have to be so re-visit these limits whenever changing infrastructure. Our servers are 32 GB, 8 core (16 thread) Xeons.

**Comment by alloylab** [comment]
have you tried this on 0.12.5?

**Comment by chuttam** [comment]
> have you tried this on 0.12.5?

Yes. Here are the open file descriptor trace results. Mac OS 10.13.4.

At times, we generate PDFs with thousands of pages and increasing the max open file descriptors is no longer an option (buffers start to overflow).

**Notes**
* We see NO problem generating large PDFs when merely passing a big HTML document to the `wkhtmltopdf` command without any `--header-html` and `--footer-html` options. We can confirm that the addition of header and/or footer is what causes the never-ending increase in open file pointers.
* Open file descriptors max out with just `--header-html`, or just `--footer-html`, or both.

```
$ wkhtmltopdf -V
wkhtmltopdf 0.12.5 (with patched qt)
```

**Command Run**

Test HTML generates ~1100 PDF pages of content when run without headers.

```
$ wkhtmltopdf -q --orientation Landscape --page-size Letter --margin-top 27 --margin-bottom 19 --header-html /Users/amit/Desktop/test-header.html /Users/amit/Desktop/test-html-content.html ./result.pdf

QEventDispatcherUNIXPrivate(): Unable to create thread pipe: Too many open files
QEventDispatcherUNIXPrivate(): Can not continue without a thread pipe
Abort trap: 6

```

**Open file pointers trace instrumentation**

Reasonable number of open file pointers get created, until such time the process starts to build the PDF outcome into `./result.pdf`. Then, a whole slew of file existence checks for the header file. Presumably, one check per page of the PDF being generated.

<img width="1120" alt="screen shot 2018-10-29 at 13 20 19" src="https://user-images.githubusercontent.com/887849/47664413-c0fcfb80-db7d-11e8-952d-0f5a69ec4070.png">


This file check goes on until the job fails when it hits the maximum number of open file descriptors/pointers defined at the OS level, which on all my systems will be smaller than the number of PDF pages being generated in one job.

<img width="1145" alt="screen shot 2018-10-29 at 13 22 28" src="https://user-images.githubusercontent.com/887849/47664629-38328f80-db7e-11e8-8099-3700810eaaad.png">

**Comment by j-prox** [comment]
> I'm coming back here two years later. We've hit some pretty big scale at our company so our servers have increased in size, including moving to bare-metal servers at http://packet.net.
> 
> We were bumping up against the `ulimit -n #=> 1024` when generating large PDFs with images and links, but we only needed to increase the limit to 2048 (`ulimit -n 2048`) and we were able to generate PDFs up to 1000 pages without a problem.
> 
> I think the size of the server plays a huge role in what these limits have to be so re-visit these limits whenever changing infrastructure. Our servers are 32 GB, 8 core (16 thread) Xeons.

Hi Josua,

I have 1GB of HTML file and it can't successfully generate it to PDF :(

**Comment by clemensutschig** [comment]
We are Blocked in this as well.. as we generate reports >150 Pages and get These errors..

**Comment by metmajer** [comment]
In line with @clemensutschig, we have now limited our uses of footer and header to text-only, as described in the mentioned issue above. This solves the issue but doesn't bring quite the same value.  Any chance to get a fix for this? Cheers!

**Comment by GuyPaddock** [comment]
Is there any reason why the files have to be open the entire time the PDF is being rendered? It seems like the `ResourceObject` and/or `MultiPageLoader` classes could load, render, and release the header and footer resources for each page instead of holding on to them. Is there a reason why this isn't the case?

**Comment by GuyPaddock** [comment]
@ashkulz can you point us toward the best maintainer to answer the design question about why the file handles need to remain open for the entire time wkhtmltopdf is operating on a given document?

**Comment by mrtronje** [comment]
@ashkulz same problem here (with current version). when generating a large pdf (900 Pages) without fooder & header we have no problems. with header and footer on hit maximum number of open file descriptors (500'000) so increasing the limit of max. open files isn't a solution.  Is there a fix in the pipeline?

**Comment by j-prox** [comment]
@mrtronje what is the file size of your HTML?

**Comment by joshuapinter** [comment]
Coming back here 5 years later after running into this again. If you look at my [previous comment](https://github.com/wkhtmltopdf/wkhtmltopdf/issues/2093#issuecomment-130106138), it was indeed correct. 

Instead of running into `Too many open files`, we were running into: 

```
[=====>                                                      ] 9%
[======>                                                     ] 10%
[======>                                                     ] 10%
*** buffer overflow detected ***: /usr/local/bin/wkhtmltopdf terminated
```

This occurred at the `Loading headers and footers (5/6)` stage.

Oddly, this wasn't reproducible on our local macOS development environments, just the Ubuntu 18 servers.

After re-checking that our limits were already set to the very generous values from my previous comment, I double checked the `header.html` content. We must have made a change to it since 2015 because now it included image URLs (to S3, if it matters) instead of referencing local image files or using base64 encoded image data.

To fix this and make it easier and DRY, we ended up adding a helper to `wicked_pdf` (the Rails gem we use that wraps `wkhtmltopdf`) that automatically reads an image URL, downloads a local copy and converts it to base64 encoded data, which is then included in the `header.html` that is passed to `wkhtmltopdf` for generating the final PDF. This immediately fixed the issue.

If you are using `wicked_pdf`, you can view and use our fork/PR of it here: https://github.com/mileszs/wicked_pdf/pull/947

If not, just double check your `header.html` and `footer.html` files and ensure there are no image URLs. Use local or base64 where possible.

Doing this, along with our increased files and stack limits, allowed us to generate 1,250 page PDFs that weighed in at over 2 GBs. 

Hope that helps others that get hit with this.

**Comment by j-prox** [comment]
> Coming back here 5 years later after running into this again. If you look at my [previous comment](https://github.com/wkhtmltopdf/wkhtmltopdf/issues/2093#issuecomment-130106138), it was indeed correct.
> 
> Instead of running into `Too many open files`, we were running into:
> 
> ```
> [=====>                                                      ] 9%
> [======>                                                     ] 10%
> [======>                                                     ] 10%
> *** buffer overflow detected ***: /usr/local/bin/wkhtmltopdf terminated
> ```
> 
> This occurred at the `Loading headers and footers (5/6)` stage.
> 
> Oddly, this wasn't reproducible on our local macOS development environments, just the Ubuntu 18 servers.
> 
> After re-checking that our limits were already set to the very generous values from my previous comment, I double checked the `header.html` content. We must have made a change to it since 2015 because now it included image URLs (to S3, if it matters) instead of referencing local image files or using base64 encoded image data.
> 
> To fix this and make it easier and DRY, we ended up adding a helper to `wicked_pdf` (the Rails gem we use that wraps `wkhtmltopdf`) that automatically reads an image URL, downloads a local copy and converts it to base64 encoded data, which is then included in the `header.html` that is passed to `wkhtmltopdf` for generating the final PDF. This immediately fixed the issue.
> 
> If you are using `wicked_pdf`, you can view and use our fork/PR of it here: [mileszs/wicked_pdf#947](https://github.com/mileszs/wicked_pdf/pull/947)
> 
> If not, just double check your `header.html` and `footer.html` files and ensure there are no image URLs. Use local or base64 where possible.
> 
> Doing this, along with our increased files and stack limits, allowed us to generate 1,250 page PDFs that weighed in at over 2 GBs.
> 
> Hope that helps others that get hit with this.

thanks for this, I will check it 👍

**Comment by jackbaron** [comment]
@joshuapinter thanks for this. but it not working.  my header.html and footer.html do not use image url. any solution for help me.

**Comment by Sicos1977** [comment]
https://github.com/Sicos1977/ChromeHtmlToPdf

**Comment by brytonbeesley** [comment]
We're generating variable-length PDF's with the company logo (size of 2 KB) in the header.  The header is the same on every page.  In scenarios where the PDF approaches 100 pages, the logo disappears.  It shows the tell-tale signs of buffer overflow.

The remedies discussed on this issue are mainly configured outside of wkhtmltopdf in the system environment.  @ashkulz, are there any options within wkhtmltopdf to directly or indirectly manage memory?  I've reviewed the documentation; just looking for confirmation that I haven't overlooked something.

Alternatively, can wkhtmltopdf be configured to create only one header file in temporary memory and reuse the same for each page when the header is the same on every page?  This would be similar to the way @natewiebe13's post on Dec 10, 2014 suggested.  The answer at the time of the post was that it would be possible after the upgrade to Qt5.  Is that a possibility now that Qt5 is supported?

Appreciate any suggestions you can give to help us solve this for our use case.  Thanks!

**Comment by ashkulz** [comment]
@brytonbeesley there aren't. Is there some reproduction for this?
